### PR TITLE
Added `rules_rust_dependencies` and `rust_register_toolchains` macros

### PR DESCRIPTION
**Pull request by UebelAndre**
This allows users to have more control over what repositories are created when setting up `rules_rust` in their workspace and conforms to the [bazel-contrib](https://github.com/bazel-contrib/rules-template/blob/f26177bcbdd9eeab373f23ca568c98704efddd3f/mylang/repositories.bzl) guide for Bazel rules.

### CODE REVIEW

File: rust/repositories.bzl [hunk 1]
```diff
@@ -40,7 +40,8 @@ def rust_repositories(
         dev_components = False,
         sha256s = None,
         include_rustc_srcs = False,
-        urls = DEFAULT_STATIC_RUST_URL_TEMPLATES):
+        urls = DEFAULT_STATIC_RUST_URL_TEMPLATES,
+        register_toolchains = True):
```
Comment (illicitonion) [inline rust/repositories.bzl:44]: I would expect this param to just bubble down to select whether `native.register_toolchains` gets run inside `rust_repository_set` - either way I would expect the repositories to exist, just not necessarily be registered as toolchains. WDYT?

**Review by illicitonion — CHANGES_REQUESTED** [review]

File: rust/repositories.bzl (hunk 1 above)
Comment (UebelAndre) [reply rust/repositories.bzl:44]: Maybe the wording is wrong but I feel like there would be little value in adding repositories to fetch toolchains but not registering them. I'd like something that prevents calling `rust_repository_set` all together unless there's a good reason to? Open to ideas

Comment (illicitonion) [reply rust/repositories.bzl:44]: I can imagine there are cases where you'd want to be able to access "a cargo for linux" binary or "a rustc for linux" binary without it affecting toolchain resolution. This is how things like [rules_foreign_cc](https://github.com/bazelbuild/rules_foreign_cc/blob/50ee9979e60e8db38e10de45d2c60873a210bf55/foreign_cc/private/framework/toolchain.bzl#L89-L111) works (and I've seen the pattern elsewhere).

Why are you calling `rust_repositories` at all if you don't want any rust repositories registered?

Comment (illicitonion) [reply rust/repositories.bzl:44]: > Though, I still want to load "rust repositories" because I want to make sure repositories used in the rules (Bazel-skylib, rules_cc, and anything else that may be added) get defined.

Sounds like we maybe could pull out a separate `rules_rust_deps` function, which `rust_repositories` would call, but which you could call instead of calling `rust_repositories`? It's all idempotent, so multiple calls should be fine...

Comment (UebelAndre) [reply rust/repositories.bzl:44]: Sure but I'd then expect to find a `register_rust_toolchains` since I don't think it's immediately clear what the difference between `rules_rust_deps` and `rust_repositories` would be.

Comment (illicitonion) [reply rust/repositories.bzl:44]: Can we bikeshed a name which makes clear "the ruleset dependencies required to use rules_rust" vs "the repositories created by rules_rust" vs "the toolchains registered for those repositories"?

https://github.com/bazel-contrib/rules-template/blob/main/mylang/repositories.bzl would suggest:

* `rules_rust_dependencies`
* `rust_toolchains`, and I'd suggest we make that with an optional `register_toolchains` boolean kwarg

Comment (UebelAndre) [reply rust/repositories.bzl:44]: But that's a larger change than I expected. Should I add a print to `rust_repositories` mentioning that the macro is deprecated?

Comment (illicitonion) [reply rust/repositories.bzl:44]: I think let's just leave `rust_repositories` as a wrapper that calls both for now, and we can consider deprecating in the future :)

Comment (UebelAndre) [reply rust/repositories.bzl:44]: Wait, but why wouldn't we deprecate this? What's the rationale behind keeping all 3 macros?

Comment (UebelAndre) [reply rust/repositories.bzl]: Also, while we're making this change, I've noticed quite a few rules use `deps.bzl` over `repositories.bzl` ([rules_go](https://github.com/bazelbuild/rules_go/blob/master/go/deps.bzl) [rules_dotnet](https://github.com/bazelbuild/rules_dotnet/blob/master/dotnet/deps.bzl), [rules_pkg](https://github.com/bazelbuild/rules_pkg/blob/main/pkg/deps.bzl)). While others do use `repositories.bzl`, I think `deps.bzl` is the more common (there are other variants too). Would this be a good time to introduce that or are we happy with `repositories.bzl`?

Comment (UebelAndre) [reply rust/repositories.bzl]: And how happy are we with the names? [rules_go](https://github.com/bazelbuild/rules_go/blob/master/go/deps.bzl#L34-L35) uses `go_rules_dependencies` and `go_register_toolchains`, I'm fine with either but want to make sure we've explored all options before making a decision.

cc @hlopko

Comment (illicitonion) [reply rust/repositories.bzl]: > Wait, but why wouldn't we deprecate this? What's the rationale behind keeping all 3 macros?

Primarily avoiding churn for people.

> Would this be a good time to introduce that or are we happy with repositories.bzl?
> And how happy are we with the names?

I'm happy with repositories.bzl, rules_rust_dependencies, and rust_register_toolchains, as it's consistent with https://github.com/bazel-contrib/rules-template but I don't have a strong opinion.

Comment (UebelAndre) [reply rust/repositories.bzl]: Updated!

**Review by illicitonion — APPROVED** [review]
Thanks!

**Review by hlopko — APPROVED** [review]
Thank you! This is also conforming to https://docs.bazel.build/versions/main/skylark/deploying.html \o/ :)

File: docs/BUILD.bazel [hunk 2]
```diff
@@ -127,12 +127,14 @@ PAGES = dict([
     page(
         name = "rust_repositories",
         symbols = [
+            "rules_rust_depednencies",
```
Comment (luiz10x) [inline docs/BUILD.bazel:130]: Drive-by comment, but should it be `rules_rust_dependencies` here?

Comment (UebelAndre) [reply docs/BUILD.bazel:130]: Oops! Thanks! https://github.com/bazelbuild/rules_rust/pull/1110
